### DISCUSSION TITLE
Unit Testing Frontend Components

### DISCUSSION THREAD

**Discussion by briangwaltney** [trpc/trpc#3612] [Q&A] [answered]
Hello,

I'm trying to figure out how to unit test a React component with Jest and Testing Library.

I apologize for being fairly new to testing in general and I'm having a really hard time figuring out how to create a TRPC context wrapper and then use it in my tests.

Ideally, I would like to mock the response data from the trpc query, but still have access to the type safety in the test during setup so my tests will fail if I change something in the router that would appropriately make the test fail.

I know I can mock all of trpc like this: 

```typescript

// Component to test
const TestComponent = () => {
  const { data: me } = trpc.user.me.useQuery();

  if (!me) return (<div>Loader</div>);

  return (<div>Ready</div>)
}


// Mock for trpc
jest.mock("@/utils/trpc", () => {
  return {
    trpc: {
      user: {
        me: {
          useQuery: jest.fn(() => {
            return jest.fn();
          }),
        },
      },
    },
  };
});

// Current unit test
test("No license required message is displayed", async () => {

  (trpc.user.me.useQuery as jest.Mock).mockReturnValueOnce({ data: user });
  render(<TestComponent />);

  const readyText = screen.queryByText(/ready/i);

  expect(readyText).toBeInTheDocument();
});
```

While this gets the test to work, it's clearly a bad solution that doesn't test any changes in my trpc implementation.

Does anyone have an example of how to test a simple component like above?

I know there is the E2E test in the example repo and I've figured out the backend unit tests using that example as well, but I would like to be able to better test individual components as needed if possible.

Any help would be greatly appreciated.

**Comment by divmgl** [comment]
Wow, thank you for the code snippet. I realize this isn't the answer you're looking for, but I think your solution for mocking tRPC is all we're going to get for the time being.

Your code snippet answers the question I asked last week: https://github.com/trpc/trpc/discussions/3670

I don't think there's much traction on providing any kind of mocking framework for tRPC/`useQuery`. This will likely need to be a grass-roots initiative done by the open-source community.

Now that I'm here and thinking about this, a potential design is to create a framework adapter (such as a Jest or VItest adapter) and pass it into a a `createMockClient` function provided by tRPC that returns all of the procedures but stubbed with `jest.fn` or `vitest.fn`. That way we can pass in a mock client into the `trpc.Provider` and use set up the mocks directly in tests. I think this would solve a majority of the use-cases with client-side testing/mocks.

**Comment by KATT** [comment]
You likely want to mount an app with a real `<TRPCProvider>` and a real server running in the test. We do this loads within our own integration testing but it's kinda hairy to look at

These are a bit messy but can have it for inspiration: 

- https://github.com/trpc/trpc/blob/2f8bcee6b6fcc2a436c1e0fa64e78eb34c0898c9/packages/tests/server/smoke.test.ts#L50-L78
- https://github.com/trpc/trpc/blob/2f8bcee6b6fcc2a436c1e0fa64e78eb34c0898c9/packages/tests/server/react/useQuery.test.tsx#L80-L117

We use `routerToServerAndClientNew` _(awful name, I know)_ to create a node server for each test.

Here's a bit of a more "standalone" implementation: https://github.com/trpc/trpc/blob/main/packages/tests/showcase/tinyrpc.test.ts

By patching some of the above together you should be able to do it. Would love to have a reference for this in the codebase or/and potentially provide utilities to make this easier if you're inclined to contribute.

**Comment by briangwaltney** [comment, marked as answer]
I actually found a solution that allowed me to test everything using a real testing backend as suggested.

Here is a repo with both front and backend tests working: https://github.com/briangwaltney/t3-testing-example.

Please note the frontend tests are silly and pointless, but the goal for this repo was to demonstrate how to create tests that work, not to produce valuable tests. In the frontend examples, the tests are effectively testing that trpc is working rather than testing any business logic in my code, but this approach can be used to mount components and hooks with trpc inside them.

This example uses a create-t3-app template, so NextAuth is included and I had to make some adjustments to get through the session creation for frontend testing.

You'll notice in the `server/auth.ts` file, I adjusted the `getServerSession` as follows:


```typescript
export const getServerAuthSession = async (ctx: {
  req: GetServerSidePropsContext["req"];
  res: GetServerSidePropsContext["res"];
}) => {
  // In test environment, we don't want to use the real next-auth session
  // because we are not going through the login flow.
  // Instead, we use the session provided by the header.
  if (process.env.APP_ENV === "test" && ctx.req.headers.session) {
    return JSON.parse(ctx.req.headers.session as string) as Session;
  }
  return await unstable_getServerSession(ctx.req, ctx.res, authOptions);
};
```
This tests for a testing environment and passes the credentials sent as a header rather than through the NextAuth cookie.

This is activated through the `utils/testWrapper.ts` functions: 

edit: included full file for clarity
```typescript
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
  createTRPCReact,
  getFetch,
  httpBatchLink,
  loggerLink,
} from "@trpc/react-query";
import superjson from "superjson";
import type { ReactElement } from "react";
import React from "react";
import type { AppRouter } from "@/server/api/root";
import fetch from "node-fetch";
import type { Session } from "next-auth";
import type { RenderOptions } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { User } from "@prisma/client";

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
const globalAny = global as any;
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
globalAny.fetch = fetch;

export const testApi = createTRPCReact<AppRouter>({
  unstable_overrides: {
    useMutation: {
      async onSuccess(opts) {
        await opts.originalFn();
        await opts.queryClient.invalidateQueries();
      },
    },
  },
});

const queryClient = new QueryClient();
const trpcClient = (session: Session | undefined) =>
  testApi.createClient({
    links: [
      loggerLink({
        enabled: (opts) => false,
        // opts.direction === "down" && opts.result instanceof Error,
      }),
      httpBatchLink({
        url: "http://localhost:3005/api/trpc",
        fetch: async (input, init?) => {
          const fetch = getFetch();
          return fetch(input, {
            ...init,
          });
        },
        ...(session ? { headers: { session: JSON.stringify(session) } } : {}),
      }),
    ],
    transformer: superjson,
  });

export function TRPCTestClientProvider(props: {
  children: React.ReactNode;
  session?: Session;
}) {
  return (
    <testApi.Provider
      client={trpcClient(props.session)}
      queryClient={queryClient}
    >
      <QueryClientProvider client={queryClient}>
        {props.children}
      </QueryClientProvider>
    </testApi.Provider>
  );
}
export const AllTheProviders: React.FC<{
  children: React.ReactNode;
  session?: Session;
}> = ({ session, children }) => {
  return (
    <TRPCTestClientProvider session={session}>
      {children}
    </TRPCTestClientProvider>
  );
};

export const createSession = (user: User): Session => ({
  user: {
    id: user.id,
  },
  expires: "2123-02-17T13:51:55.373Z",
});

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & {
    user?: User;
  }
) => {
  const session = options?.user ? createSession(options.user) : undefined;

  return render(ui, {
    wrapper: (props) => <AllTheProviders {...props} session={session} />,
    ...options,
  });
};

export const hookWrapper = (user?: User) =>
  function wrapperOptions(props: { children: React.ReactNode }) {
    const session = user ? createSession(user) : undefined;
    return <AllTheProviders {...props} session={session} />;
  };

export * from "@testing-library/react";
export { customRender as render };
```
When the wrapper is used in a hook or component render, and a `user` is passed into the wrapper, the session is created as a header and sent to the test backend like this:

```typescript
  test("should render", async () => {
    const { user } = await createUser();

    render(<ProtectedRoute />, { user: user });

    expect(screen.getByText("Loading...")).toBeInTheDocument();

    const message = await screen.findByText(/data/i);

    expect(message).toBeInTheDocument();
  });
```

This approach has worked well for me so far and allows me to modify or reset the db before every test if desired although that seems to lead to run into some race issues with vitest concurrency and disabling the threads doesn't seem to fix it, so I typically just reset the test db periodically rather than before every test.

I've not contributed to a project like this before, so I'm not sure what would be most helpful from this example repo or how best to add it, but I'm happy to do a more in-depth explanation of the code and the process if it would be helpful. I just need a little guidance on what you would like to see.

**Comment by getinnocuous** [comment]
I managed to create something similar by using [msw-trpc](https://github.com/maloguertin/msw-trpc) and then creating a custom render function with RTL. Posting here just in case it helps someone achieve what I was trying to.

```
import { AppRouter } from '../server/routers/_app'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import type { RenderOptions } from '@testing-library/react'
import { render } from '@testing-library/react'
import { createTRPCReact, httpBatchLink, loggerLink } from '@trpc/react-query'
import { createTRPCMsw } from 'msw-trpc'
import fetch from 'node-fetch'
import React, { ReactElement } from 'react'

// @ts-expect-error
global.fetch = fetch

const mockedtRPC = createTRPCReact<AppRouter>({
  unstable_overrides: {
    useMutation: {
      async onSuccess(opts) {
        await opts.originalFn()
        await opts.queryClient.invalidateQueries()
      },
    },
  },
})

const mockedtRPCClient = mockedtRPC.createClient({
  links: [
    loggerLink({
      enabled: () => false,
    }),
    httpBatchLink({
      url: `http://localhost:${process.env.PORT ?? 3000}/api/trpc`,
    }),
  ],
})

const mockedQueryClient = new QueryClient()

const TRPClientProvider = (props: { children: React.ReactNode }) => {
  return (
    <mockedtRPC.Provider
      client={mockedtRPCClient}
      queryClient={mockedQueryClient}
    >
      <QueryClientProvider client={mockedQueryClient}>
        {props.children}
      </QueryClientProvider>
    </mockedtRPC.Provider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => {
  return render(ui, {
    wrapper: (props) => <TRPClientProvider {...props} />,
    ...options,
  })
}

export const trpcMsw = createTRPCMsw<AppRouter>()
export * from '@testing-library/react'
export { customRender as render }
```

**Reply by CUexter** [reply]
@getinnocuous can you give me a more detailed example, I can't seem to get pass this issue
https://github.com/maloguertin/msw-trpc/issues/6

**Reply by getinnocuous** [reply]
I've tried it using your eslint file and it also get the same error. I think this is an issue with how your eslint is configured.

This eslint setup works for me 

```
{
  "extends": [
    "next/core-web-vitals",
    "prettier",
    "plugin:@typescript-eslint/recommended"
  ],
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "rules": {
    "no-unused-vars": ["warn", { "argsIgnorePattern": "^_" }],
    "@typescript-eslint/no-unused-vars": [
      "warn",
      { "argsIgnorePattern": "^_" }
    ],
    "@typescript-eslint/ban-ts-comment": "off",
    "@typescript-eslint/no-empty-function": "off",
    "react/jsx-curly-brace-presence": [
      1,
      {
        "props": "always",
        "children": "never",
        "propElementValues": "always"
      }
    ]
  }
}
```

**Reply by CUexter** [reply]
Turning off no-unsafe-argument will work but i think there should be a better way. Also how did you set up the mocking server ? My handlers doesn't work

**Reply by jharlow** [reply]
Awesome work! Cleared an error I was really struggling with. The only thing stopping me from making a couple of pages of documentation based on this answer is that my mocked mutations and queries using `trpcMsw` always return undefined. Even more bizarrely, I'm even receiving requests in my MSW mocks but nothing is being returned!

Going to give it another bit or work but does that issue bring anything to mind?

> **Update**: I've figured out my problem, I think that it's because my `transformer` configuration for the my `trpcMsw` setup. @getinnocuous This is exactly the solution I was searching for. Would you mind if I tried to find somewhere to write some documentation for this solution, crediting you? Again, super impressed. _Thank you_ 🎆
